### PR TITLE
Bpbrowser 584

### PR DESCRIPTION
**Pull request by scribe**
This change prevents us from sending a dead symlink, but allows the job to continue.
It also prevents us from sending a job that is nothing but dead symlinks

### CODE REVIEW

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 1]
```diff
@@ -117,6 +125,15 @@ data class PutJobData(
         val parent = item.second.parent
         val paths = item.second.toFile().walk(FileWalkDirection.TOP_DOWN)
                     .filter { !(it.isDirectory && Files.list(it.toPath()).use { f -> f.findAny().isPresent }) }
+                    .filter {
+                     try {
+                        it.toPath().toRealPath()
+                        true
+                    } catch (e: java.nio.file.NoSuchFileException) {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: Why did you specify the absolute class name here?  Shouldn't the import at the top eliminate the need for this?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 2]
```diff
@@ -117,6 +125,15 @@ data class PutJobData(
         val parent = item.second.parent
         val paths = item.second.toFile().walk(FileWalkDirection.TOP_DOWN)
                     .filter { !(it.isDirectory && Files.list(it.toPath()).use { f -> f.findAny().isPresent }) }
+                    .filter {
+                     try {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: I would pull this logic out into a function to clean up the function call chain.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt (hunk 1 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: When I use just the class name it ends up with the kotlin.io.NoSuchFileException which doesn't match what's thrown.

Is there a better way?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt (hunk 2 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: That's a good idea I'll get that swapped around.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt (hunk 1 above)
Comment (rpmoore) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: Explicitly specify the import at the top.  If you don't you'll get the kotlin one.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 3]
```diff
@@ -112,20 +120,34 @@ data class PutJobData(
...
-                            targetDir + parent.relativize(it.toPath()).toString().replace(localDelim, "/") + "/"
-                        } else {
-                            targetDir + parent.relativize(it.toPath()).toString().replace(localDelim, "/")
-                        }, if (Files.isDirectory(it.toPath())) 0L else it.length())
-                    }
-                .map { Pair(it, item.second) }
-        return paths
+    private fun pathToDs3Object(path: Path): Sequence<Pair<Ds3Object, Path>> {
+        return path.toFile().walk(FileWalkDirection.TOP_DOWN)
+            .filter(::rejectEmptyDirectory)
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: This is way cleaner.  It really communicates the intention behind what each of the steps is doing.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 4]
```diff
@@ -112,20 +120,34 @@ data class PutJobData(
...
+    }
+
+    private fun rejectEmptyDirectory(file: File) =
+        !(file.isDirectory && Files.list(file.toPath()).use { f -> f.findAny().isPresent })
+
+    private fun convertToBPDelimiter(file: File): Ds3Object {
+        val parent = file.toPath().parent
+        val localDelim = file.toPath().fileSystem.separator
+        val size = if (Files.isDirectory(file.toPath())) 0L else file.length()
+        val basePath: String = targetDir + parent.relativize(file.toPath()).toString().replace(localDelim, "/") +
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]:
```suggestion
        val basePathBuilder = StringBuilder(targetDir)
            .append(parent.relativize(file.toPath()).toString().replace(localDelim, "/"))
        if (file.isDirectory) basePathBuilder.append("/")
        
        return Ds3Object(basePath.toString(), size)
```

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 5]
```diff
@@ -112,20 +120,34 @@ data class PutJobData(
...
-                        Ds3Object(if (it.isDirectory) {
-                            targetDir + parent.relativize(it.toPath()).toString().replace(localDelim, "/") + "/"
-                        } else {
-                            targetDir + parent.relativize(it.toPath()).toString().replace(localDelim, "/")
-                        }, if (Files.isDirectory(it.toPath())) 0L else it.length())
-                    }
-                .map { Pair(it, item.second) }
-        return paths
+    private fun pathToDs3Object(path: Path): Sequence<Pair<Ds3Object, Path>> {
+        return path.toFile().walk(FileWalkDirection.TOP_DOWN)
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: Move `.walk` to it's own line

**Pull request merged**